website/teacher.py

```diff
@@ -253,6 +258,63 @@ def update_level_preferences(user, class_id):
             return {'achievement': achievement}, 200
         return {}, 200
 
+    @app.route('/invite_student', methods=['POST'])
+    @requires_login
+    def invite_student(user):
+        body = request.json
+        # Validations
+        if not isinstance(body, dict):
+            return g.auth_texts.get('ajax_error'), 400
+        if not isinstance(body.get('username'), str):
+            return g.auth_texts.get('username_invalid'), 400
+        if not isinstance(body.get('class_id'), str):
+            return 'class id must be a string', 400
+
+        username = body.get('username')
+        class_id = body.get('class_id')
+
+        if not is_teacher(user):
+            return utils.error_page(error=403, ui_message='retrieve_class')
+        Class = DATABASE.get_class(class_id)
+        if not Class or Class['teacher'] != user['username']:
+            return utils.error_page(error=404, ui_message='no_such_class')
+
+        user = DATABASE.user_by_username(username)
+        if not user:
+            return g.auth_texts.get('student_not_existing'), 400
+        if 'students' in Class and user['username'] in Class['students']:
+            return g.auth_texts.get('student_already_in_class'), 400
+        if DATABASE.get_username_invite(user['username']):
+            return g.auth_texts.get('student_already_invite'), 400
+
+        # So: The class and student exist and are currently not a combination -> invite!
+        DATABASE.add_class_invite(username, class_id)
+        return {}, 200
+
+    @app.route('/remove_student_invite', methods=['POST'])
+    @requires_login
+    def remove_invite(user):
+        body = request.json
+        # Validations
+        if not isinstance(body, dict):
+            return 'body must be an object', 400
```

Comment:
Good point! However the error throw in this cause should never happen: The username and class id are extracted from the table row at the point the "remove" button is pressed. So they should always be a string, the user is unable to "screw this up". They are only here to make sure the package sent is valid (and give ourselves some useful debugging information if necessary). But still, I agree that it would be best to return a more useful error message to the user in the unexpected case this happens. Will fix!

Reply:
Fixed!